### ISSUE TITLE
Unbounded loops in message-lane

### ISSUE THREAD

**Issue by svyatonik** [paritytech/parity-bridges-common#539]
Before starting working on weights in message-lane module, we need to have upper bound on all loops in message-lane module. Right now there are four loops in message-lane calls:
- [x] unbounded [messages dispatch loop](https://github.com/paritytech/parity-bridges-common/blob/18a4e98d05b94974ea3da49c03994f4b155d3cc3/modules/message-lane/src/lib.rs#L346) in `fn receive_messages_proof()`. It is unbounded, because we only limit messages by dispatch weight (and dispatch weight may be zero). This makes impossible to compute base weight of delivery transaction. The base weight should not be larger than `TargetChain.MaximalExtrinsicWeight` - [`MaximalMessageWeight`](https://github.com/paritytech/parity-bridges-common/blob/18a4e98d05b94974ea3da49c03994f4b155d3cc3/bin/runtime-common/src/messages.rs#L50), so that we always can deliver single large message. At the same time, we want more than one message in the batch, so it shouldn't be minimal;
- [x] unbounded [relayers reward loop](https://github.com/paritytech/parity-bridges-common/blob/18a4e98d05b94974ea3da49c03994f4b155d3cc3/modules/message-lane/src/lib.rs#L402). It was assumed to be bounded by the [`MaxUnconfirmedMessagesAtInboundLane`](https://github.com/paritytech/parity-bridges-common/blob/18a4e98d05b94974ea3da49c03994f4b155d3cc3/modules/message-lane/src/lib.rs#L84) parameter at the bridged chain, but after #418 it is no longer true. There are several ways to fix this:
 - treat `MaxUnconfirmedMessagesAtInboundLane` as it was designed initially. The optimization may stay, but only as storage optimization (i.e. instead of having N entries for N messages relayed by single relayer, we may have 1 entry for N messages);
 - add another 'strict' upper bound;
 - add some structure that will optimize payouts computation (so we won't need to read N messages to pay reward for these messages);
 - introduce non-instant rewards - rewards may be paid e.g. when message is being pruned (in block initialization or when new message is being sent);

- [x] bounded [prune messages loop](https://github.com/paritytech/parity-bridges-common/blob/18a4e98d05b94974ea3da49c03994f4b155d3cc3/modules/message-lane/src/outbound_lane.rs#L90);
- [x] bounded [relayers list maintain loop](https://github.com/paritytech/parity-bridges-common/blob/18a4e98d05b94974ea3da49c03994f4b155d3cc3/modules/message-lane/src/inbound_lane.rs#L66).

**Comment by tomusdrw** [comment]
>  dispatch weight may be zero

How come the dispatch weight might be zero? If that's the case let's introduce a "minimal dispatch cost" (kind of like "base extrinsict cost") and verify it on the source side like in #536 maybe?

> unbounded relayers reward loop. 

The upper loop is still bounded, it's the [inner (nonce loop)](https://github.com/paritytech/parity-bridges-common/blob/18a4e98d05b94974ea3da49c03994f4b155d3cc3/modules/message-lane/src/lib.rs#L406-L416) being unbonded. I've checked the code and it can actually be greately optimized, cause instead of calling `pay_relayers_reward` for each nonce separately, we should sum up the `message_data.fee` and just call it once.
Given we have this optimisation and the fact that upper loop is bounded, I guess the only thing is to limit the range of `nonces`, we can simply introduce a limit that matches say 10x limit of transactions of target-chain dispatch (maximal).
So for instance if the relayer can deliver `100` messages at once (in a single transaction). The limit here should be `10 * 100`. That gives us a theoretical upper bound of the entire loop as `10 * 100 * MaxUnconfirmedMessagesAtInboundLane`. So if we have a single relayer, she will be able to collect rewards for up to 10 dispatches, but with multiple relayers they would rather be bound by `MaxUnconfirmedMessagesAtInboundLane`, not this.

The limit should also be enforced on the target chain - i.e. when we reach `100` nonces by a single relayer, we should split this into another storage entry, cause otherwise it would be impossible for them to collect rewards. For safety off-limit rewards collections should not fail, but rather we should ignore them or process up to the upper bound (i.e. even if the proof contains a delivery of 120 messages, you only get rewarded for the first 100).

**Comment by svyatonik** [comment]
> How come the dispatch weight might be zero? If that's the case let's introduce a "minimal dispatch cost" (kind of like "base extrinsict cost") and verify it on the source side like in #536 maybe?

I can be zero, because in #536 we now only verify that per-byte weight/fee is paid. So sending empty payload (i.e. `vec![]`) with declared dispatch weight = zero is acceptable. I don't think introducing any minimal weight is necessary - you could find many calls with weight set to zero in substrate. I don't think we should introduce additional barriers for messages.

> Given we have this optimisation and the fact that upper loop is bounded, I guess the only thing is to limit the range of nonces, we can simply introduce a limit that matches say 10x limit of transactions of target-chain dispatch (maximal).

Range of nonces is exactly what has been bounded initially with `MaxUnconfirmedMessagesAtInboundLane` (this has also bounded number of rewarded relayers). Now we don't have this - that's what needs to be fixed now. I can't say I understand your idea - I'm not sure about the introduced connection between delivery transaction and confirmation transaction (why would we need 10 * 100 - messages may be delivered one-by-one, or in batches, why would we need to have any assumptions about that here?). But I guess it is something similar to what I meant under "add another 'strict' upper bound" (another constant that would limit number of messages we need to read in rewards loop). Anyway - right now I think having single constant instead of two is a way to go, but I haven't decided yet. Will get back to this after dealing with dispatch loop.

**Comment by tomusdrw** [comment]
> zero is acceptable

Hmm, yeah, you are totally right. I find it quite interesting though, obviously sending `vec![]` as `Call` would fail the dispatch, but it still for instance creates an event on the target chain and requires all the bridge machinery to confirm delivery, etc. I wonder if we shouldn't charge some extra fixed cost for every message.

>  I can't say I understand your idea [...] but I guess it is something similar to what I meant under "add another 'strict' upper bound"

Yes, I meant to introduce an upper bound for the nonce range exactly, but consider it separately from `MaxUnconfirmed..`.

> why would we need 10 * 100 - messages may be delivered one-by-one, or in batches, why would we need to have any assumptions about that here?

So I was thinking about what we can optimize here. The simplest case is where for each delivery transaction the relayer sends a confirm-delivery transaction on the source chain. However I think we should make it possible to batch confirmations, since (in average case) should be cheaper than dispatch transactions. So I imagined a case where we have a single relayer, submitting a maximal number of messages in every transaction and then confirming them with an order of magnitude less transactions (obviously the `10` factor is up for the discussion; that's `10 * 100` iterations of the loop, but only 1 transaction). That's the "happy case". The "unhappy case" is when we have multiple relayers each delivering only one message per transaction (that's one iteration of the loop, but many transactions). Without extra incentives they would rather only confirm their own deliveries, so we end up with no optimization at all.
The worst case is obviously when we have multiple relayers trying to attack the chain by delivering many transactions in a batch and then not confirming them, so to unstuck the lane we need to confirm all of them hence `10 * 100 * MaxUnconfirmed...`.

We can solve the "unhappy" case by introducing incentives, we obviously need to care about the worst case anyway, and since I have an optimistic nature, I wanted to enable the "happy" variant as well, cause I imagine for many lanes there would only be one layer (especially initially, without a more sophisticated incentivisation mechanism (see https://github.com/paritytech/parity-bridges-common/issues/316#issuecomment-679015022))

**Comment by svyatonik** [comment]
> Hmm, yeah, you are totally right. I find it quite interesting though, obviously sending vec![] as Call would fail the dispatch, but it still for instance creates an event on the target chain and requires all the bridge machinery to confirm delivery, etc. I wonder if we shouldn't charge some extra fixed cost for every message.

The fee that the sender is paying consists of: message dispatch fee (that's what this zero weight is about), base delivery transaction fee (which I guess would be the max weight of delivery transaction with single message), base confirmation transaction cost (... same here ...), base reward confirmation cost (... same here ...) and relayer interest (atm it is configured to be configured to be 10% of all the previous fees sum). So the message sender will, no doubt, pay more than for the same transaction (with call set to `vec![]`) on the target chain. Please remember that the dispatch event is included in delivery transaction cost, relay operational cost is included in that relayer fee, everything else about confirmation is included in the confirmation transaction cost. So I don't think something else should be added to this cost.

> So I was thinking about what we can optimize here.

Ok, I think now I got your idea about two consts.
